light_curves/scale_up.md

```diff
@@ -0,0 +1,558 @@
+- The script tells the user what the process PIDs are and where the log and data files are.
+- In case the run needs to be canceled, the script can find and kill all the processes it launched.
+- The script can monitor `top` during the run, saving `top` output to a log file at a user-defined interval(s). The helper can be used to load this file in python.
+
++++
+
+### Parallel processing methods: bash script vs. python's `multiprocessing`
+
++++
+
+- Bash script. Recommended for most runs with medium to large samples. Allows ZTF to use additional parallelization internally, and so is often faster (ZTF often takes the longest and returns the most data for AGN-like samples). Writes stdout and stderr to log files, useful for monitoring jobs and resource usage. Can monitor and record `top` to help identify CPU and RAM usage/needs.
+- Python's `multiprocessing` library. Useful as a demonstration. May be convenient for runs with small to medium sample sizes. Has drawbacks that may be significant including the inability to use ZTF's internal parallelization and that it does not save the log output (stdout and stderr) to file.
+
++++
+
+### Imports
+
+```{code-cell}
+# Ensure all dependencies are installed
+!pip install -r requirements.txt
+```
+
+```{code-cell}
+import json  # create json strings for bash script arguments
+import multiprocessing  # python parallelization method
+import pandas as pd  # use a DataFrame to work with light-curve and other data
+import sys  # add code directories to the path
+
+sys.path.append("code_src/")
+import helpers.scale_up  # python "helper" for parallelization and large scale runs
+import helpers.top  # load `top` output to DataFrames and make figures
+from data_structures import MultiIndexDFObject  # load light curve data as a MultiIndexDFObject
+from plot_functions import create_figures  # make light curve figures
+```
+
+## Example 1: Multi-wavelength light curves for 500,000 SDSS AGN
+
++++
+
+This example shows how to launch a large-scale run using the bash script, monitor its performance, and diagnose a problem (out of RAM).
+This run collects light curves for 500,000 SDSS objects and takes several hours to complete, but is not actually executed here.
+Instead, we show the bash commands and then look at logs that were generated by running the commands on 2024/03/01.
+
++++
+
+### Launch the run
+
++++
+
+```bash
+# This will launch a very large scale run.
+# See Example 3 for a detailed explanation of parameter options.
+# If you are running this for the first time, reduce the sample size 500000 -> 50.
+
+# run_id will be used in future calls to manage this run, and also determines the name of the output directory.
+$ run_id="demo-SDSS-500k"
+
+# Execute the run:
+$ bash code_src/helpers/scale_up.sh \
+    -r "$run_id" \
+    -j '{"get_samples": {"SDSS": {"num": 500000}}, "archives": {"ZTF": {"nworkers": 8}}}' \
+    -a "core"  # shortcut for "Gaia HEASARC IceCube WISE ZTF"
+```
+
```

Comment:
clarify if this run is only for ZTF, since ZTF is explicitly mentioned, but then the line below it implies that it is searching Gaia, Heasarc, Icecube, WISE, ZTF.  

Also, one day we will need to change "core" to include MAST, or just remove it.

Reply:
Clarification added. 

Also, "core" and "all" were useful shortcuts for me so that I didn't have to remember them every time. But if you'd prefer not to expose them here in the notebook we could just list out the archive names.

Reply:
I replaced "core" in all the function and script calls with the explicit list of archives. I also scrubbed the reference out of all the documentation. The actual list still exists and can be used, but the user would have to look directly at the `ARCHIVE_NAMES` dictionary where it's defined in order to find it. I also I renamed it to "scaled" since this is the subset that is expected to run successfully at scale. Probably still not the best name, but better than "core".